src/rmarkdown/preview.ts

```diff
@@ -0,0 +1,328 @@
+
+import * as cp from 'child_process';
+import * as vscode from 'vscode';
+import * as fs from 'fs-extra';
+import * as os from 'os';
+
+import { doWithProgress, getRpath, readContent, setContext } from '../util';
+import { extensionContext } from '../extension';
+import path = require('path');
+
+class RMarkdownChild extends vscode.Disposable {
+    title: string;
+    cp: cp.ChildProcessWithoutNullStreams;
+    panel: vscode.WebviewPanel;
+    resourceViewColumn: vscode.ViewColumn;
+    uri: vscode.Uri;
+    outputUri: vscode.Uri;
+    dir: string;
+    fileWatcher: fs.FSWatcher;
+
+    constructor(title: string, cp: cp.ChildProcessWithoutNullStreams, panel: vscode.WebviewPanel, resourceViewColumn: vscode.ViewColumn, outputUri: vscode.Uri, uri: vscode.Uri, dir: string) {
+        super(() => {
+            this.cp?.kill('SIGKILL');
+            this.panel?.dispose();
+            this.fileWatcher?.close();
+            fs.removeSync(this.outputUri.path);
+        });
+
+        this.title = title;
+        this.cp = cp;
+        this.panel = panel;
+        this.resourceViewColumn = resourceViewColumn;
+        this.uri = uri;
+        this.outputUri = outputUri;
+        this.dir = dir;
+    }
+
+    startFileWatcher(RMarkdownPreviewManager: RMarkdownPreviewManager) {
+        let fsTimeout: NodeJS.Timeout;
+        const docWatcher = fs.watch(this.uri.path, {}, () => {
+            if (!fsTimeout) {
+                fsTimeout = setTimeout(() => { fsTimeout = null; }, 1000);
+                void RMarkdownPreviewManager.updatePreview(this);
+            }
+        });
+        this.fileWatcher = docWatcher;
+    }
+}
+
+class RMarkdownChildStore extends vscode.Disposable {
+    private store: Set<RMarkdownChild> = new Set<RMarkdownChild>();
```

Comment:
Is there a reason we don't use a `Map` here?

Reply:
Not particularly - were you thinking something like `Map<vscode.Uri, RMarkdownChild>();`, where the uri is the input path?

Reply:
Yes, exactly.